File: app/src/main/java/eu/kanade/tachiyomi/data/sync/GoogleDriveTransport.kt
package eu.kanade.tachiyomi.data.sync

import eu.kanade.tachiyomi.network.NetworkHelper
import eu.kanade.tachiyomi.network.await
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.asRequestBody
import okio.BufferedSink
import okio.source
import org.json.JSONObject
import java.io.File

/**
 * Stores a single .tachibk file on Google Drive using an OAuth 2.0 access token (e.g. one minted
 * via the OAuth Playground for the scope `https://www.googleapis.com/auth/drive.file`). Merge
 * happens client-side via BackupRestorer.
 *
 * The access token is reused from the sync apiKey field. The user is responsible for refreshing
 * it; this transport does not perform the OAuth refresh dance.
 */
class GoogleDriveTransport(
    private val accessToken: String,
    private val networkHelper: NetworkHelper,
) : SyncTransport.FileStorage {

    override suspend fun pull(): ByteArray? {
        val fileId = lookupFileId() ?: return null

        val downloadUrl = "$DRIVE_API/files/$fileId".toHttpUrl().newBuilder()
            .addQueryParameter("alt", "media")
            .build()
        val request = Request.Builder()
            .url(downloadUrl)
            .header("Authorization", "Bearer $accessToken")
            .get()
            .build()

        val response = networkHelper.client.newCall(request).await()
        try {
            return when (response.code) {
                in 200..299 -> response.body.bytes().takeIf { it.isNotEmpty() }
                404 -> null
                401, 403 -> throw SyncException("Drive auth failed (HTTP ${response.code})")
                else -> throw SyncException("Drive pull failed (HTTP ${response.code})")
            }
        } finally {
            response.close()
        }
    }

    override suspend fun push(payload: File) {
        val existing = lookupFileId()

        val request = if (existing == null) {
            // Multipart create. The payload is streamed directly from disk inside the multipart
            // body so we never load the backup into memory.
            val boundary = "mihon_sync_${System.currentTimeMillis()}"
            val metadata = JSONObject().apply { put("name", FILE_NAME) }.toString()
            Request.Builder()
                .url("$UPLOAD_API/files?uploadType=multipart")
                .header("Authorization", "Bearer $accessToken")
                .post(MultipartRelatedBody(boundary, metadata, payload))
                .build()
        } else {
            Request.Builder()
                .url("$UPLOAD_API/files/$existing?uploadType=media")
                .header("Authorization", "Bearer $accessToken")
                .patch(payload.asRequestBody(OCTET_STREAM))
                .build()
        }

        val response = networkHelper.client.newCall(request).await()
        try {
            if (!response.isSuccessful) {
                throw SyncException("Drive push failed (HTTP ${response.code})")
            }
        } finally {
            response.close()
        }
    }

    private suspend fun lookupFileId(): String? {
        val url = "$DRIVE_API/files".toHttpUrl().newBuilder()
            .addQueryParameter("q", "name='$FILE_NAME' and trashed=false")
            .addQueryParameter("spaces", "drive")
            .addQueryParameter("fields", "files(id,name)")
            .addQueryParameter("pageSize", "1")
            .build()

        val request = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $accessToken")
            .get()
            .build()

        val response = networkHelper.client.newCall(request).await()
        try {
            if (!response.isSuccessful) {
                if (response.code == 401 || response.code == 403) {
                    throw SyncException("Drive auth failed (HTTP ${response.code})")
                }
                throw SyncException("Drive lookup failed (HTTP ${response.code})")
            }
            val json = JSONObject(response.body.string())
            val files = json.optJSONArray("files") ?: return null
            if (files.length() == 0) return null
            return files.getJSONObject(0).optString("id").takeIf { it.isNotBlank() }
        } finally {
            response.close()
        }
    }

    companion object {
        private const val FILE_NAME = "mihon-sync.tachibk"
        private const val DRIVE_API = "https://www.googleapis.com/drive/v3"
        private const val UPLOAD_API = "https://www.googleapis.com/upload/drive/v3"
        private val OCTET_STREAM = "application/octet-stream".toMediaTypeOrNull()
    }
}

/**
 * Streaming multipart/related body for Drive's create-file endpoint. Writes the metadata part,
 * then the payload file (chunked off disk via Okio), then the trailing boundary — never
 * materialising the file in memory.
 */
private class MultipartRelatedBody(
    boundary: String,
    metadataJson: String,
    private val payload: File,
) : RequestBody() {

    private val prologue: ByteArray = buildString {
        append("--").append(boundary).append("\r\n")
        append("Content-Type: application/json; charset=UTF-8\r\n\r\n")
        append(metadataJson).append("\r\n")
        append("--").append(boundary).append("\r\n")
        append("Content-Type: application/octet-stream\r\n\r\n")
    }.toByteArray(Charsets.UTF_8)

    private val epilogue: ByteArray = "\r\n--$boundary--".toByteArray(Charsets.UTF_8)

    private val mediaType: MediaType? =
        "multipart/related; boundary=$boundary".toMediaTypeOrNull()

    override fun contentType(): MediaType? = mediaType

    override fun contentLength(): Long =
        prologue.size.toLong() + payload.length() + epilogue.size.toLong()

    override fun writeTo(sink: BufferedSink) {
        sink.write(prologue)
        payload.inputStream().use { input ->
            input.source().use { source ->
                sink.writeAll(source)
            }
        }
        sink.write(epilogue)
    }
}

File: flutter/lib/data/sync/google_drive_transport.dart
/// Google Drive transport. Stores a single .tachibk file in the user's
/// Drive root. Auth uses a bearer access token minted by the user (e.g.
/// via the OAuth Playground with scope
/// `https://www.googleapis.com/auth/drive.file`). Token refresh is the
/// user's problem for v1.0 — same as Mihon.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:dio/dio.dart';

import 'sync_transport.dart';

class GoogleDriveTransport extends SyncTransportFileStorage {
  GoogleDriveTransport({required this.accessToken, required this.dio});

  final String accessToken;
  final Dio dio;

  static const String _driveApi = 'https://www.googleapis.com/drive/v3';
  static const String _uploadApi =
      'https://www.googleapis.com/upload/drive/v3';
  static const String _fileName = 'mohyeong-sync.tachibk';

  @override
  Future<Uint8List?> pull() async {
    final fileId = await _lookupFileId();
    if (fileId == null) return null;
    try {
      final response = await dio.get<List<int>>(
        '$_driveApi/files/$fileId',
        queryParameters: {'alt': 'media'},
        options: Options(
          responseType: ResponseType.bytes,
          headers: {'Authorization': 'Bearer $accessToken'},
          validateStatus: (s) =>
              s != null && (s == 404 || (s >= 200 && s < 300)),
        ),
      );
      if (response.statusCode == 404) return null;
      final bytes = response.data;
      if (bytes == null || bytes.isEmpty) return null;
      return Uint8List.fromList(bytes);
    } on DioException catch (e) {
      final code = e.response?.statusCode;
      if (code == 401 || code == 403) {
        throw SyncException('Drive auth failed (HTTP $code)');
      }
      throw SyncException('Drive pull failed (HTTP ${code ?? '?'})');
    }
  }

  @override
  Future<void> push(Uint8List payload) async {
    final existing = await _lookupFileId();
    try {
      if (existing == null) {
        // Multipart create. Mihon streams off disk via OkHttp; we just
        // build a single bytes payload because Dart's HTTP libs handle
        // megabyte-sized buffers cheaply and library snapshots stay
        // well under that ceiling in practice.
        final boundary =
            'mohyeong_sync_${DateTime.now().microsecondsSinceEpoch}';
        final metadata = jsonEncode({'name': _fileName});
        final builder = BytesBuilder(copy: false)
          ..add(utf8.encode('--$boundary\r\n'
              'Content-Type: application/json; charset=UTF-8\r\n\r\n'
              '$metadata\r\n'
              '--$boundary\r\n'
              'Content-Type: application/octet-stream\r\n\r\n'))
          ..add(payload)
          ..add(utf8.encode('\r\n--$boundary--'));
        final body = builder.toBytes();

        await dio.post<void>(
          '$_uploadApi/files',
          queryParameters: {'uploadType': 'multipart'},
          data: Stream.value(body),
          options: Options(
            headers: {
              'Authorization': 'Bearer $accessToken',
              'Content-Type': 'multipart/related; boundary=$boundary',
              'Content-Length': body.length,
            },
            validateStatus: (s) => s != null && s >= 200 && s < 300,
          ),
        );
      } else {
        await dio.patch<void>(
          '$_uploadApi/files/$existing',
          queryParameters: {'uploadType': 'media'},
          data: Stream.value(payload),
          options: Options(
            headers: {
              'Authorization': 'Bearer $accessToken',
              'Content-Type': 'application/octet-stream',
              'Content-Length': payload.length,
            },
            validateStatus: (s) => s != null && s >= 200 && s < 300,
          ),
        );
      }
    } on DioException catch (e) {
      throw SyncException(
          'Drive push failed (HTTP ${e.response?.statusCode ?? '?'})');
    }
  }

  Future<String?> _lookupFileId() async {
    try {
      final response = await dio.get<Map<String, dynamic>>(
        '$_driveApi/files',
        queryParameters: {
          'q': "name='$_fileName' and trashed=false",
          'spaces': 'drive',
          'fields': 'files(id,name)',
          'pageSize': '1',
        },
        options: Options(
          headers: {'Authorization': 'Bearer $accessToken'},
          validateStatus: (s) => s != null && s >= 200 && s < 300,
        ),
      );
      final files = response.data?['files'] as List<dynamic>?;
      if (files == null || files.isEmpty) return null;
      final id = (files.first as Map<String, dynamic>)['id'] as String?;
      return (id == null || id.isEmpty) ? null : id;
    } on DioException catch (e) {
      final code = e.response?.statusCode;
      if (code == 401 || code == 403) {
        throw SyncException('Drive auth failed (HTTP $code)');
      }
      throw SyncException('Drive lookup failed (HTTP ${code ?? '?'})');
    }
  }
}
